File: PP/Lista1/palindrome.ml
(*
    O(n)
    a' list -> bool
*)
let palindrome xs =
    let rec reverse xss =
        match xss with
        | [] -> []
        | h :: t -> reverse t @ [h]
    in if reverse xs = xs then true else false;;

let int_of_bool b = if b then 1 else 0;;

let rec print_list xs =
    match xs with
    | [] -> ()
    | e::l -> print_char e ; print_string " " ; print_list l;;

print_int (int_of_bool (palindrome ['a';'l';'a']));;

File: PP/Lista1/palindrome.scala
object palindrome {
    def main(args: Array[String]) : Unit = {
        def palindrome[A](xs : List[A]) : Boolean =
            def reverse[A](xss : List[A]) : List[A] =
                xss match {
                    case List() => List()
                    case h :: t => List.concat(reverse(t), List(h))
                }
            if reverse(xs) == xs then true else false
        println(palindrome(List('a','l','i')))
    }
}